pyrdp/player/gdi/draw.py

```diff
@@ -0,0 +1,702 @@
+    StreamBitmapFirst, StreamBitmapNext, GdiPlusFirst, GdiPlusNext, GdiPlusEnd, GdiPlusCacheFirst, \
+    GdiPlusCacheNext, GdiPlusCacheEnd, FrameMarker
+from pyrdp.parser.rdp.orders.secondary import CacheBitmapV1, CacheBitmapV2, CacheBitmapV3, CacheColorTable, \
+    CacheGlyph, CacheBrush
+from pyrdp.parser.rdp.orders.primary import DstBlt, PatBlt, ScrBlt, DrawNineGrid, MultiDrawNineGrid, \
+    LineTo, OpaqueRect, SaveBitmap, MemBlt, Mem3Blt, MultiDstBlt, MultiPatBlt, MultiScrBlt, MultiOpaqueRect, \
+    FastIndex, PolygonSc, PolygonCb, PolyLine, FastGlyph, EllipseSc, EllipseCb, GlyphIndex, Brush, \
+    BACKMODE_TRANSPARENT
+
+from pyrdp.ui import QRemoteDesktop, RDPBitmapToQtImage
+
+from .cache import BitmapCache, BrushCache, PaletteCache, GlyphCache, GlyphEntry
+from .raster import set_rop3, set_rop2
+
+from PySide2.QtCore import Qt, QPoint
+from PySide2.QtGui import QImage, QPainter, QColor, QPixmap, QBrush, QPen, QPolygon
+
+LOG = logging.getLogger(__name__)
+
+SCREEN_BITMAP_SURFACE = 0xFFFF
+BITMAPCACHE_SCREEN_ID = 0xFF
+SUBSTITUTE_SURFACE = -1
+
+# Opaque Rectangle Encoding Flags.
+OPRECT_BOTTOM_ABSENT = 0x01
+OPRECT_RIGHT_ABSENT = 0x02
+OPRECT_TOP_ABSENT = 0x04
+OPRECT_LEFT_ABSENT = 0x08
+
+GLYPH_SPECIAL_PROCESSING = -32768
+GLYPH_FRAGMENT_USE = 0xFE
+GLYPH_FRAGMENT_ADD = 0xFF
+
+# Defined in 2.2.2.2.1.1.2.13 (GlyphIndex)
+SO_VERTICAL = 0x01
+SO_HORIZONTAL = 0x02
+SO_CHAR_INC_EQUAL_BM_BASE = 0x20
+
+
+def rgb_to_qcolor(color: int):
+    """Convert an RDP color (0xRRGGBB) to a QColor."""
+    bpp = 16
+    # TODO: check BPP from capabilities
+    if bpp in [24, 32]:
+        return QColor(color)
+    if bpp == 16:
+        t = color & 0x1F
+        t = (t << 3) + t // 4
+        b = min(t, 255)
+
+        t = (color >> 5) & 0x3F
+        t = (t << 2) + t // 4 // 2
+        g = min(t, 255)
+
+        t = (color >> 11) & 0x1F
+        t = (t << 3) + t // 4
+        r = min(t, 255)
+
+    elif bpp == 15:
+        pass
+    elif bpp == 8:  # TODO: Support palettized mode.
+        pass
+    return QColor.fromRgb(r, g, b)
+
+
+class GdiQtFrontend(GdiFrontend):
+    """
+    A Qt Frontend for GDI drawing operations.
+
+    This acts as a straight adapter from GDI to Qt as much as
+    possible, but GDI specific operations that are not supported by Qt
+    are implemented here.
+
+    Some of the methods are not implemented and will simply ignore the
+    order:
+
+        - NineGrid-related orders
+        - StreamBitmap orders
+        - GDI+ orders
+
+    Note that `CacheBitmapV3` is only used in 32 bits-per-pixel which
+    PyRDP currently does not support due to RDP6.0 compression not
+    being implemented yet.
+
+    If you have a replay which contains any unsupported or untested
+    order, do not hesitate to share it with the project maintainers so
+    that support can be added as required.
+
+    (Make sure that the trace does not contain any sensitive information)
```

Comment:
This should be added to the README as well. It is too important to be only here.